---
title: "Murine Cerebellum scRNA-seq Analysis"
output: html_document
date: "2025-05-13"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("Seurat")
library("tidyverse")
library("Matrix")
library("stringr")
library("dplyr")
library("ggplot2")
library("reshape2")
library("purrr")
library("hdf5r")
library("remotes")
library("loupeR")
```


```{r, echo=TRUE, results='hide', message=FALSE}

# Preparing folder format for Seurat Object creation

raw_dir <- "GSE224466_RAW"
all_files <- list.files(raw_dir, full.names = TRUE, recursive = FALSE)

# Extract unique sample prefixes

prefixes <- unique(gsub("_(barcodes|genes|matrix)\\.tsv$|_matrix\\.mtx$", "", basename(all_files)))


# Create sub-folders and move files

for (prefix in prefixes) {
  sample_dir <- file.path(raw_dir, prefix)
  dir.create(sample_dir, showWarnings = FALSE)

  files_to_move <- list.files(
    raw_dir, 
    pattern = paste0("^", prefix, ".*\\.(tsv|mtx)$"), full.names = TRUE)
    
  file.rename(files_to_move,
            file.path(sample_dir, basename(files_to_move)))}
```



```{r }

# Reading each sample and creating a Seurat list with 24 objects

seurat_list <- lapply(all_files, function(dir) {
  sample_name <- basename(dir)
  data <- Read10X(data.dir = dir)
  seurat_obj <- CreateSeuratObject(counts = data, project = sample_name)
  seurat_obj$sample <- sample_name
  return(seurat_obj)})

# Exploring the data structure

seurat_list[[1]]@meta.data
```

```{r}
# Checking to see if the sample names were applied correctly 

lapply(seurat_list, function(x) x$sample[1])

# Number of cells with 0 RNA counts for each sample before filtering 

sapply(seurat_list, function(seurat_obj) {sum(seurat_obj$nCount_RNA == 0, na.rm = TRUE)})

# 12,13 and 14 have a very high percentage of cells with 0 UMI counts

```

```{r}
# Identifying mitochondrial and ribosomal protein genes

genes_file <- "GSE224466_RAW/GSM7024833_PAN_CB_E13_A/genes.tsv"
genes_df <- read.table(genes_file, header = FALSE, col.names = c("ensembl_id", "gene_symbol"))
mito_genes_symbols <- grep("^mt-", genes_df$gene_symbol, value = TRUE, ignore.case = TRUE)
mito_ensembl_ids <- genes_df$ensembl_id[genes_df$gene_symbol %in% mito_genes_symbols]

print(mito_genes_symbols)

seurat_list <- lapply(seurat_list, function(seurat_obj) {
  seurat_obj[["percent.mt"]] <- PercentageFeatureSet(seurat_obj, features = mito_genes_symbols)
  return(seurat_obj)
})

head(seurat_list[[1]][["percent.mt"]])



ribo_genes_symbols <- grep("^Rp[sl]", genes_df$gene_symbol, value = TRUE, ignore.case = TRUE)
ribo_genes_symbols <- ribo_genes_symbols[!grepl("ka|kc|-ps", ribo_genes_symbols)]
ribo_ensembl_ids <- genes_df$ensembl_id[genes_df$gene_symbol %in% ribo_genes_symbols]

print(ribo_genes_symbols)

seurat_list <- lapply(seurat_list, function(seurat_obj) {
  seurat_obj[["percent.ribo"]] <- PercentageFeatureSet(seurat_obj, features = ribo_genes_symbols)
  return(seurat_obj)
})

head(seurat_list[[1]][["percent.ribo"]])

```
``` {r}
# Sub-setting the each object to filter out genes with over 10 % mitochondrial DNA and less than 3500 or 2000 unique transcripts and over 15,000 transcripts. 

filtered_list <- lapply(seurat_list, function (seurat_obj) {
  seurat_obj <- subset(seurat_obj, subset =  nCount_RNA > 3500 & nCount_RNA < 15000)
  return (seurat_obj)
})


saveRDS(filtered_list, file = "filtered_list.rds")
# filtered_list <- readRDS("filtered_list.rds"

filtered_list2 <- lapply(seurat_list, function (seurat_obj) {
  seurat_obj <- subset(seurat_obj, subset =  nCount_RNA > 2750 & nCount_RNA < 15000)
  return (seurat_obj)
})

saveRDS(filtered_list2, file = "filtered_list2.rds")
#filtered_list2 <- readRDS("filtered_list2.rds")


# Once this is applied, majority of the cells are already below the threshold of 10 percent, but I will filter to be sure.  

filtered_list <- lapply(filtered_list, function (seurat_obj) {
  seurat_obj <- subset(seurat_obj, subset = percent.mt < 10)
  return (seurat_obj)
})

filtered_list2 <- lapply(filtered_list2, function (seurat_obj) {
  seurat_obj <- subset(seurat_obj, subset = percent.mt < 10)
  return (seurat_obj)
})


# Summary statistics for each object 

lapply(filtered_list, function(seurat_obj) {summary(seurat_obj$nCount_RNA)})
lapply(filtered_list2, function(seurat_obj) {summary(seurat_obj$nCount_RNA)})

# Count of cells after filtering

for (i in seq_along(filtered_list)) {cat("Sample", i, "has", ncol(filtered_list[[i]]), "cells after filtering\n")}

for (i in seq_along(filtered_list2)) {cat("Sample", i, "has", ncol(filtered_list2[[i]]), "cells after filtering\n")}

# Mean nCount_RNA after filtering
sapply(filtered_list, function(obj) {tapply(obj$nCount_RNA, obj$sample, mean)})
sapply(filtered_list2, function(obj) {tapply(obj$nCount_RNA, obj$sample, mean)})


```


``` {r}
# Violin plot for % of mitochondrial genes in the first sample, checking to confirm that no cells have MT content above the threshold. 

VlnPlot(filtered_list[[1]], features = "percent.mt", pt.size = 0.1, layer = "counts") + NoLegend()

VlnPlot(filtered_list[[1]], features = "percent.ribo", pt.size = 0.1, layer = "counts") + NoLegend()

```



``` {r}
# Removing mitochondrial & ribosomal genes

genes_to_remove <- unique(c(mito_genes_symbols, ribo_genes_symbols))

filtered_list <- lapply(filtered_list, function(seurat_obj) {
  genes_keep <- setdiff(rownames(seurat_obj), genes_to_remove)
  subset(seurat_obj, features = genes_keep)
})

filtered_list2 <- lapply(filtered_list2, function(seurat_obj) {
  genes_keep <- setdiff(rownames(seurat_obj), genes_to_remove)
  subset(seurat_obj, features = genes_keep)
})
  
# Removing genes with 0 counts across all cells 
  
filtered_list <- lapply(filtered_list, function(obj) {
  counts <- GetAssayData(obj, slot = "counts")
  genes_to_keep <- rowSums(counts) > 0
  subset(obj, features = names(genes_to_keep[genes_to_keep]))
})


filtered_list2 <- lapply(filtered_list2, function(obj) {
  counts <- GetAssayData(obj, slot = "counts")
  genes_to_keep <- rowSums(counts) > 0
  subset(obj, features = names(genes_to_keep[genes_to_keep]))
})

```

``` {r}
# Additional summary statistics

sapply(filtered_list, ncol)

sum(sapply(filtered_list, ncol))

sapply(seurat_list, function(seurat_obj) {sum(seurat_obj$nCount_RNA == 0, na.rm = TRUE)})

sapply(seurat_list, function(seurat_obj) {sum(seurat_obj$nCount_RNA > 2000 & seurat_obj$nCount_RNA < 15000, na.rm = TRUE)})

```



``` {r}

sample_names <- sapply(filtered_list, function(x) unique(x$sample))
names(filtered_list) <- sample_names
names(filtered_list2) <- sample_names

summary_df <- data.frame(
  Sample = names(filtered_list2),
  Avg_nFeature_RNA = sapply(filtered_list, function(obj) mean(obj@meta.data$nFeature_RNA)),
  Avg_nCount_RNA = sapply(filtered_list, function(obj) mean(obj@meta.data$nCount_RNA))
)

print(summary_df)

melted_df <- melt(summary_df, id.vars = "Sample")

# Plot both metrics side-by-side
ggplot(melted_df, aes(x = Sample, y = value, fill = variable)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(x = "Sample", y = "Average", fill = "") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ggtitle("Average nFeature_RNA and nCount_RNA per Sample")


```





``` {r}
# Merging all objects into one Seurat object 

filtered_list <- lapply(names(filtered_list), function(name) {
  RenameCells(filtered_list[[name]], add.cell.id = name)
})
names(filtered_list) <- sample_names
merged_obj <- reduce(filtered_list, merge)


filtered_list2 <- lapply(names(filtered_list2), function(name) {
  RenameCells(filtered_list2[[name]], add.cell.id = name)
})
names(filtered_list2) <- sample_names
merged_obj2 <- reduce(filtered_list2, merge)

```


``` {r}
# Normalizing the data 

merged_obj <- NormalizeData(merged_obj, verbose = FALSE)
merged_obj <- FindVariableFeatures(merged_obj, selection.method = "vst", nfeatures = 2000, verbose = FALSE) # Based on the 2000 most variable features
merged_obj <- ScaleData(merged_obj, vars.to.regress = c("percent.mt","nCount_RNA"), verbose = FALSE)


merged_obj2 <- NormalizeData(merged_obj2, verbose = FALSE)
merged_obj2 <- FindVariableFeatures(merged_obj2, selection.method = "vst", nfeatures = 2000, verbose = FALSE) # Based on the 2000 most variable features
merged_obj2 <- ScaleData(merged_obj2, vars.to.regress = c("percent.mt","nCount_RNA"), verbose = FALSE)

``` 


``` {r}
# PCA, clustering and UMAP

merged_obj <- RunPCA(merged_obj, features = VariableFeatures(merged_obj), npcs = 40, verbose = FALSE)
merged_obj <- RunUMAP(merged_obj, dims = 1:40, verbose = FALSE)
merged_obj <- FindNeighbors(merged_obj, dims = 1:40, verbose = FALSE)
merged_obj <- FindClusters(merged_obj, resolution = 0.5, verbose = FALSE)

saveRDS(merged_obj, file = "merged_obj.rds")
merged_obj <- readRDS("merged_obj.rds")

merged_obj2 <- RunPCA(merged_obj2, features = VariableFeatures(merged_obj2), npcs = 40, verbose = FALSE)
merged_obj2 <- RunUMAP(merged_obj2, dims = 1:40, verbose = FALSE)
merged_obj2 <- FindNeighbors(merged_obj2, dims = 1:40, verbose = FALSE)
merged_obj2 <- FindClusters(merged_obj2, resolution = 0.5, verbose = FALSE)

saveRDS(merged_obj2, file = "merged_obj2.rds")
merged_obj2 <- readRDS("merged_obj2.rds")

# Cluster summary
table(merged_obj$seurat_clusters)
table(merged_obj2$seurat_clusters)

```


``` {r}
# Checking # of PCAs to use for UMAP

ElbowPlot(merged_obj, ndims = 40)
ElbowPlot(merged_obj2, ndims = 40)

#Visualization of UMAP

DimPlot(merged_obj, reduction = "umap", label = TRUE, group.by = "seurat_clusters")
DimPlot(merged_obj2, reduction = "umap", label = TRUE, group.by = "seurat_clusters")


# DoHeatmap(obj, features = unique(top_markers[["E12"]]$gene))

``` 






``` {r}
# Determination and Annotation of marker genes for each cluster

merged_obj <- JoinLayers(merged_obj)
all_markers <- FindAllMarkers(merged_obj, assay ="RNA", slot = "data", only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)


merged_obj2 <- JoinLayers(merged_obj2)
all_markers2 <- FindAllMarkers(merged_obj2, assay ="RNA", slot = "data", only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)


names(all_markers)

top_markers <- all_markers %>%
    group_by(cluster) %>%
    slice_max(order_by = avg_log2FC, n = 10) %>%
    ungroup()

top_markers2 <- all_markers2 %>%
    group_by(cluster) %>%
    slice_max(order_by = avg_log2FC, n = 10) %>%
    ungroup()


```

```{r}
# Manual annotation of each cluster
cluster_ids <- c(
  "0" = "Midbrain Glutamatergic Cells", "1" = "", "2" = "", "3" = "Bergmann Glia", "4" = "",
  "5" = "GABAergic Progenitors", "6" = "VZ? Progenitor Cells" ,"7" = "Glutamatergic CNs",
  "8" = "GABAergic Interneurons", "9" = "Early Progenitor Cells","10" = "Roof Plate Cells",
  "11" = "Radial Glial Cells", "12" = "Glutamatergic CNs", "13" = "Early Progenitor Cells", "14" = "Erythrocytes","15" = "GABAergic Interneurons","16" = "Endothelial Cell Precursors",
  "17" = "Microglia", "18" = "Mesodermal Cells?", "19" = "GABAergic Interneurons?",
  "20" = "Pericytes","21" = "Erythrocytes","22" = "Oligodendrocytes",
  "23" = "Ciliated Cells")

merged_obj <- RenameIdents(merged_obj, cluster_ids)



cluster_ids2 <- c(
  "0" = "Neural progenitor",
  "1" = "Radial glia",
  "2" = "Interneurons",
  "3" = "Astrocytes",
  "4" = "Excitatory neurons",
  "5" = "Oligodendrocyte precursors",
  ...
)
merged_obj <- RenameIdents(merged_obj, cluster_ids)

# check cell labelin in Loupe and also CellxFate, see any differences
```





``` {r}
# Creating Loupe objects for further analysis

merged_obj_1 <- JoinLayers(merged_obj)

create_loupe_from_seurat(merged_obj_1, output_name = "loupe_1")

merged_obj_2 <- JoinLayers(merged_obj2)

create_loupe_from_seurat(merged_obj_2, output_name = "loupe_2")


options(Seurat.object.assay.version = "v3")
options(Seurat.object.assay.version = "v4")
```


